<a href="https://colab.research.google.com/github/maykelsilva300/Lenguaje_de_programacion/blob/main/Ejercio_Resuelto_VPN_Maykel.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Caso práctico: Analista de inversiones

Imagina que trabajas como **analista financiero** en una empresa de inversión.

Tu equipo debe decidir en cuál proyecto invertir.

La empresa solo puede elegir **un proyecto**.

Tu tarea será usar **Python para analizar la rentabilidad de cada proyecto** utilizando el **Valor Presente Neto (VPN)**.

## ¿Qué es el Valor Presente Neto (VPN)?

El **Valor Presente Neto (VPN)** permite determinar si un proyecto de inversión es rentable.

- Si **VPN > 0** → el proyecto genera valor.
- Si **VPN < 0** → el proyecto no es rentable.

Fórmula:

$$VPN = \\sum \
rac{Flujo_t}{(1+r)^t} - Inversión\ Inicial$$

## Proyectos de inversión

Usaremos **diccionarios en Python** para representar cada proyecto.

In [ ]:
proyecto_A = {
    'nombre': 'Proyecto Solar',
    'inversion': 8000,
    'flujos': [2500, 3000, 3500, 4000],
    'tasa': 0.10
}

proyecto_B = {
    'nombre': 'Proyecto Tecnológico',
    'inversion': 10000,
    'flujos': [3000, 3500, 4200, 4500],
    'tasa': 0.10
}

proyecto_C = {
    'nombre': 'Proyecto Logístico',
    'inversion': 7000,
    'flujos': [1800, 2600, 3200, 3800],
    'tasa': 0.10
}

## Calcular el VPN usando `for`

In [39]:
vpna = -proyecto_A['inversion']

for t, flujo in enumerate(proyecto_A['flujos']):
    vpna += flujo / (1 + proyecto_A['tasa'])**(t+1)

print('VPN del proyecto A:', round(vpna, 3))

VPN del proyecto A: 2113.722


## Ejercicio

Calcule el VPN del **Proyecto B** y **Proyecto C**.

In [36]:
vpnb = -proyecto_B['inversion']

for t, flujo in enumerate(proyecto_B['flujos']):
    vpnb += flujo / (1 + proyecto_B['tasa'])**(t+1)

print('VPN del proyecto B:', round(vpnb, 3))


VPN del proyecto B: 1848.917


In [37]:
vpnc = -proyecto_C['inversion']

for t, flujo in enumerate(proyecto_C['flujos']):
    vpnc += flujo / (1 + proyecto_C['tasa'])**(t+1)

print('VPN del proyecto C:', round(vpnc, 3))

VPN del proyecto C: 1784.782


## Decisión con `if`

In [46]:
vpns = [vpna, vpnb, vpnc]
for vpn in vpns:
    if vpn > 0:
        print('El proyecto con el valor del vpn igual', vpn, 'es rentable')
    else:
        print('El proyecto con el valor del vpn igual', vpn, 'No es rentable')
Mayor = max(vpns)
print('El Proyecto que tiene la VPN más alto es:',Mayor)

El proyecto con el valor del vpn igual 2113.7217403182817 es rentable
El proyecto con el valor del vpn igual 1848.9174236732433 es rentable
El proyecto con el valor del vpn igual 1784.7824602144638 es rentable
El Proyecto que tiene la VPN más alto es: 2113.7217403182817


## Analizar todos los proyectos

In [ ]:
proyectos = [proyecto_A, proyecto_B, proyecto_C]

for proyecto in proyectos:
    vpn = -proyecto['inversion']

    for t, flujo in enumerate(proyecto['flujos']):
        vpn += flujo / (1 + proyecto['tasa'])**(t+1)

    print(proyecto['nombre'], 'VPN:', round(vpn, 2))

Proyecto Solar VPN: 2113.72
Proyecto Tecnológico VPN: 1848.92
Proyecto Logístico VPN: 1784.78


## Simulación de tasas con `while`

In [ ]:
tasa = 0.05

while tasa <= 0.20:
    vpn = -proyecto_A['inversion']

    for t, flujo in enumerate(proyecto_A['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

    print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: 3416.28
Tasa: 0.1 VPN: 2113.72
Tasa: 0.15000000000000002 VPN: 1030.66
Tasa: 0.2 VPN: 121.14


In [29]:
tasa = 0.05

while tasa <= 0.20:
    vpn = -proyecto_B['inversion']

    for t, flujo in enumerate(proyecto_B['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

    print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: 3362.03
Tasa: 0.1 VPN: 1848.92
Tasa: 0.15000000000000002 VPN: 589.66
Tasa: 0.2 VPN: -468.75


In [30]:
tasa = 0.05

while tasa <= 0.20:
    vpn = -proyecto_C['inversion']

    for t, flujo in enumerate(proyecto_C['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

    print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: 2963.11
Tasa: 0.1 VPN: 1784.78
Tasa: 0.15000000000000002 VPN: 807.91
Tasa: 0.2 VPN: -10.03


## Reto final

1. Calcule el VPN de los tres proyectos.
2. Determine cuál proyecto tiene el **VPN más alto**.
3. Justifique la decisión.


1.R/ Se realiza el calculo del VPN de los tres proyectos donde nos arroja los siguientes datos:

Proyecto Solar VPN: 2113.72
Proyecto Tecnológico VPN: 1848.92
Proyecto Logístico VPN: 1784.78

2.R/ Al realizar el analisis podemos confirmar que el proyecto con el VPN más alto es el Proyecto Solar con 2113.72

3.R/ Desde un punto de vista estrictamente financiero, el Proyecto Solar es la decisión correcta porque maximiza la creación de valor para los inversionistas. El VPN representa la riqueza adicional que el proyecto genera hoy, después de cubrir la inversión inicial y los costos de capital. Un VPN de 2,113.72 significa que, tras recuperar cada peso invertido y pagar la tasa de interés exigida, el proyecto le añade ese valor neto a la empresa.


